packages/cli/src/rpc/explain/navie/history.ts

```diff
@@ -0,0 +1,350 @@
+      await writeFile(projectDirectoriesFile, projectDirectories.join('\n'));
+    }
+    return threadDir;
+  }
+
+  static threadAccessError(threadId: string, action: string, e: any): ThreadAccessError {
+    const err = e as Error & { code?: string };
+    if (err.code === 'ENOENT') warn(`Thread ${threadId} not found`);
+
+    return new ThreadAccessError(threadId, action, e instanceof Error ? e : undefined);
+  }
+
+  // In the old-style history format, threads are stored in files named by their thread id
+  // and timestamp. In the new-style history format, threads are stored in files named by
+  // their thread id, and the thread file is symlinked to a file named for the date of the thread.
+  static async migrate(
+    oldDirectory: string,
+    history: History,
+    options: { cleanup: boolean } = { cleanup: true }
+  ): Promise<void> {
+    // List the contents of the old directory. Each one is a threadId directory, containing
+    // timestamped messages.
+    const threadIds = (await readdir(oldDirectory)).filter(
+      // Match UUIDs like 5278527e-c4ed-4e45-9fb6-372ed6a036f6 in the thread dir
+      // Being careful not to touch anything else, like the dates and threads directories that are
+      // created in the new history format.
+      (dir) => dir.match(THREAD_ID_REGEX)
+    );
+
+    const threadFileAsTimestamp = (threadFile: string) => parseInt(threadFile.split('.')[0]);
+
+    for (const threadId of threadIds) {
+      warn(`[history] Migrating thread ${threadId} from old history format`);
+
+      const threadDir = join(oldDirectory, threadId);
+      const threadFiles = await readdir(threadDir);
+      threadFiles.sort((a, b) => threadFileAsTimestamp(a) - threadFileAsTimestamp(b));
+
+      let lastUserMessageId: string | undefined;
+      for (const threadFile of threadFiles) {
+        const timestamp = threadFileAsTimestamp(threadFile);
+        // MessageIds are not available, so fake them with timestamps.
+        const messageId = timestamp.toString();
+        const messageFile = join(threadDir, threadFile);
+        const messageStr = await readFile(messageFile, 'utf-8');
+        const message = JSON.parse(messageStr) as Message;
+        if (message.role === 'user') {
+          lastUserMessageId = messageId;
+          // codeSelection and prompt are not available in the old format.
+          await history.question(threadId, messageId, message.content, undefined, undefined);
+        } else if (message.role === 'assistant') {
+          if (lastUserMessageId)
+            await history.token(threadId, lastUserMessageId, messageId, message.content);
+        }
+
+        // Project directories are unknown, so overwrite the file contents with empty text.
+        const projectDirectoriesFile = join(
+          history.directory,
+          'threads',
+          threadId,
+          'projectDirectories.txt'
+        );
+        await writeFile(projectDirectoriesFile, '');
```

Comment:
I'm seeing this locally.

```
3399894 [Stderr] [history] Migrating thread 1866418e-0b40-4393-95f2-003b45ebf23e from old history format
3399894 [Stderr] Error: ENOENT: no such file or directory, open '/home/db/.appmap/navie/history/threads/1866418e-0b40-4393-95f2-003b45ebf23e/projectDirectories.txt'
3399894 [Stderr]     at async open (node:internal/fs/promises:639:25)
3399894 [Stderr]     at async writeFile (node:internal/fs/promises:1219:14)
3399894 [Stderr]     at async History.migrate (/home/db/dev/applandinc/appmap-js/packages/cli/built/rpc/explain/navie/history.js:271:17)
3399894 [Stderr]     at async Object.handler (/home/db/dev/applandinc/appmap-js/packages/cli/built/cmds/index/rpc.js:76:9) {
3399894 [Stderr]   errno: -2,
3399894 [Stderr]   code: 'ENOENT',
3399894 [Stderr]   syscall: 'open',
3399894 [Stderr]   path: '/home/db/.appmap/navie/history/threads/1866418e-0b40-4393-95f2-003b45ebf23e/projectDirectories.txt'
3399894 [Stderr] }
3399894 [Stderr] exited with code 1
3399894 [Stderr] too many crashes - aborting
```

Reply:
I've added an update for this, it should ignore old history that contains no files or unrecognized roles.

Check thread `1866418e-0b40-4393-95f2-003b45ebf23e` to see if it's either:

a) Empty
b) Does not contain any messages that have either the `user` or `assistant` role.